core/src/kura.rs

```diff
@@ -4,258 +4,504 @@
-    sync::Arc,
+    sync::{Arc, Mutex, Weak},
 };
 
-use async_trait::async_trait;
-use futures::{Stream, StreamExt, TryStreamExt};
-use iroha_actor::{broker::*, prelude::*};
-use iroha_crypto::{HashOf, MerkleTree};
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use iroha_actor::broker::*;
+use iroha_crypto::HashOf;
 use iroha_logger::prelude::*;
-use iroha_version::{
-    scale::{DecodeVersioned, EncodeVersioned},
-    try_decode_all_or_just_decode,
-};
-use pin_project::pin_project;
+use iroha_version::scale::{DecodeVersioned, EncodeVersioned};
 use serde::{Deserialize, Serialize};
-use tokio::{
-    fs,
-    io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt, BufReader, BufWriter},
-};
-use tokio_stream::wrappers::ReadDirStream;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
 
 use crate::{block::VersionedCommittedBlock, block_sync::ContinueSync, prelude::*, sumeragi};
 
-/// Message for storing committed block
-#[derive(Clone, Debug, Message)]
-pub struct StoreBlock(pub VersionedCommittedBlock);
-
-/// Gets hash of some specific block by height
-#[derive(Clone, Copy, Debug, Message)]
-#[message(result = "Option<HashOf<VersionedCommittedBlock>>")]
-pub struct GetBlockHash {
-    /// The block's height
-    pub height: usize,
-}
-
-/// High level data storage representation.
-/// Provides all necessary methods to read and write data, hides implementation details.
-#[derive(Debug)]
-pub struct KuraWithIO<IO> {
+/// The interface of Kura subsystem
+pub struct Kura {
     // TODO: Kura doesn't have different initialisation modes!!!
     #[allow(dead_code)]
     mode: Mode,
-    block_store: BlockStore<IO>,
-    merkle_tree: MerkleTree<VersionedCommittedBlock>,
+    block_store: Mutex<Box<dyn BlockStoreTrait + Send>>,
+    block_hash_array: Mutex<Vec<HashOf<VersionedCommittedBlock>>>,
     wsv: Arc<WorldStateView>,
     broker: Broker,
-    actor_channel_capacity: u32,
+    block_reciever: Mutex<Receiver<VersionedCommittedBlock>>,
+    block_sender: Sender<VersionedCommittedBlock>,
 }
 
-/// Production qualification of `KuraWithIO`
-pub type Kura = KuraWithIO<DefaultIO>;
-
-/// Generic implementation for tests - accepting IO mocks
-impl<IO: DiskIO> KuraWithIO<IO> {
-    /// ctor
+impl Kura {
+    /// Initialize Kura and start a thread that recieves
+    /// and stores new blocks.
     /// # Errors
-    /// Will forward error from `BlockStore` construction
-    pub async fn new_meta(
+    /// Fails if there are filesystem errors when trying
+    /// to access the block store indicated by the provided
+    /// path.
+    pub fn new(
         mode: Mode,
         block_store_path: &Path,
-        blocks_per_file: NonZeroU64,
         wsv: Arc<WorldStateView>,
         broker: Broker,
-        actor_channel_capacity: u32,
-        io: IO,
-    ) -> Result<Self> {
-        Ok(Self {
+        block_channel_size: u32,
+    ) -> Result<Arc<Self>> {
+        let (block_sender, block_reciever) = channel(block_channel_size as usize);
```

Comment:
This as well cannot fail, _ever_, so there is no point in using `try_into`.

Reply:
We have already discussed that some months ago. The idea is that it looks like a C-style cast. If you have any assumption like *"it will never fail"* then you can show this assumption in the code using `expect`

Reply:
Consider that someone will refactor the type of  `block_channel_size` to `u64` using the auto-refactoring feature in the IDE. Then this `as` cast will be incorrect, but Rust won't say anything to you.
Using `expect` we can find this error quicker, but still only in runtime

Reply:
You make a good point. I will change it.